File: new-resource-detector/src/test/java/nl/haploid/resource/detector/EventConsumerServiceIT.java
package nl.haploid.resource.detector;

import nl.haploid.resource.detector.service.EventConsumerService;
import org.apache.kafka.clients.producer.KafkaProducer;
import org.apache.kafka.clients.producer.ProducerRecord;
import org.junit.Assert;
import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.Timeout;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;

import java.util.UUID;
import java.util.concurrent.ExecutionException;

public class EventConsumerServiceIT extends AbstractIT {

    @Autowired
    private EventConsumerService service;

    @Autowired
    private KafkaProducer<String, String> kafkaProducer;

    @Value("${kafka.topic}")
    private String topic;

    @Rule
    public Timeout globalTimeout = new Timeout(10000);

    @Test
    public void testConsumeSingleEvent() throws InterruptedException, ExecutionException {
        final String expectedMessage = UUID.randomUUID().toString();
        final ProducerRecord<String, String> record = new ProducerRecord<String, String>(topic, expectedMessage);
        kafkaProducer.send(record).get();
        final String actualMessage = service.consumeSingleEvent();
        Assert.assertEquals(expectedMessage, actualMessage);
    }
}

File: event-consumer/src/test/scala/nl/haploid/octowight/consumer/service/EventConsumerServiceIT.scala
package nl.haploid.octowight.consumer.service

import java.util.concurrent.TimeUnit

import nl.haploid.octowight.consumer.{AbstractIT, TestData}
import nl.haploid.octowight.kafka.producer.KafkaOutChannel
import nl.haploid.octowight.{AtomChangeEvent, JsonMapper}
import org.apache.kafka.clients.producer.ProducerRecord
import org.junit.Rule
import org.junit.rules.Timeout
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.test.util.ReflectionTestUtils

class EventConsumerServiceIT extends AbstractIT with KafkaOutChannel {
  private[this] lazy val log = LoggerFactory.getLogger(getClass)

  @Autowired private[this] val eventConsumerService1: EventConsumerService = null
  @Autowired private[this] val eventConsumerService2: EventConsumerService = null
  @Autowired private[this] val jsonMapper: JsonMapper = null

  @Rule val globalTimeout = new Timeout(10, TimeUnit.SECONDS)

  private[this] val topic = TestData.nextString

  behavior of "Event consumer service"

  override def beforeEach() = {
    log.info(s"Topic is $topic")
    super.beforeEach()
    ReflectionTestUtils.setField(eventConsumerService1, "topic", topic)
  }

  private[this] def sendMessage(topic: String, event: AtomChangeEvent): Unit = {
    val message = jsonMapper.serialize(event)
    log.debug(s"Send message: $message")
    val record = new ProducerRecord[String, String](topic, message)
    kafkaProducer.send(record).get
  }

  it should "consume a message" in {
    val expectedEvent = TestData.atomChangeEvent(TestData.nextString)
    sendMessage(topic, expectedEvent)
    val actualEvent = eventConsumerService1.consumeEvent
    actualEvent should be(expectedEvent)
  }

  it should "consume multiple messages" in {
    val event1 = TestData.atomChangeEvent(TestData.nextString)
    val event2 = TestData.atomChangeEvent(TestData.nextString)
    val expectedEvents = Set(event1, event2)
    sendMessage(topic, event1)
    sendMessage(topic, event2)
    val actualEvents = eventConsumerService1.consumeDistinctEvents()
    actualEvents should be(expectedEvents)
  }

  it should "commit stream offset" in {
    val event = TestData.atomChangeEvent("harris")
    sendMessage(topic, event)
    eventConsumerService1.consumeDistinctEvents()
    eventConsumerService1.commit()
    val actualEvents = eventConsumerService2.consumeDistinctEvents()
    actualEvents should have size 0
  }
}
